homeassistant/components/ring/__init__.py

```diff
@@ -148,13 +141,103 @@ def hub_refresh():
     if not unload_ok:
         return False
 
-    await hass.async_add_executor_job(hass.data[DATA_TRACK_INTERVAL])
+    hass.data[DOMAIN].pop(entry.entry_id)
+
+    if len(hass.data[DOMAIN]) != 0:
+        return True
 
+    # Last entry unloaded, clean up
+    hass.data.pop(DATA_TRACK_INTERVAL)()
+    hass.data.pop(DATA_HEALTH_DATA_TRACKER)
+    hass.data.pop(DATA_HISTORY)
     hass.services.async_remove(DOMAIN, "update")
 
-    hass.data.pop(DATA_RING_DOORBELLS)
-    hass.data.pop(DATA_RING_STICKUP_CAMS)
-    hass.data.pop(DATA_RING_CHIMES)
-    hass.data.pop(DATA_TRACK_INTERVAL)
+    return True
+
+
+class HealthDataUpdater:
+    """Data storage for health data."""
+
+    def __init__(self, hass):
+        """Track devices that need health data updated."""
+        self.hass = hass
+        self.devices = {}
+        self._unsub_interval = None
+
+    async def track_device(self, config_entry_id, device):
+        """Track a device."""
+        if not self.devices:
+            self._unsub_interval = async_track_time_interval(
+                self.hass, self.refresh_all, SCAN_INTERVAL
+            )
+
+        key = (config_entry_id, device.device_id)
+
+        if key not in self.devices:
+            self.devices[key] = {
+                "device": device,
+                "count": 1,
+            }
+        else:
+            self.devices[key]["count"] += 1
+
+        await self.hass.async_add_executor_job(device.update_health_data)
+
+    @callback
+    def untrack_device(self, config_entry_id, device):
+        """Untrack a device."""
+        key = (config_entry_id, device.device_id)
+        self.devices[key]["count"] -= 1
+
+        if self.devices[key]["count"] == 0:
+            self.devices.pop(key)
+
+        if not self.devices:
+            self._unsub_interval()
+            self._unsub_interval = None
+
+    def refresh_all(self, _):
+        """Refresh all registered devices."""
+        for info in self.devices.values():
+            info["device"].update_health_data()
+
+        dispatcher_send(self.hass, SIGNAL_UPDATE_HEALTH_RING)
+
+
+class HistoryCache:
+    """Helper to fetch history."""
+
+    STALE_AFTER = 10  # seconds
+
+    def __init__(self, hass):
+        """Initialize history cache."""
+        self.hass = hass
+        self.cache = {}
+
+    async def async_get_history(self, config_entry_id, device):
```

Comment:
Is this better than using a dict cache and our `Throttle` helper to throttle updating the cache?

Reply:
Yeah, because with the `Throttle` decorator, if a call is in process, it will return `None` instead of waiting for the call to finish.

Reply:
So at worst we would then get 10 second stale data?

What kind of data is this?

Reply:
Last 10 events. 

Throttle returns `None`, not the data. The implementation would become then `data.update(); handle(data.data)`

Reply:
I think that if this approach seems to work, we might want to turn that into a generic asyncio throttle function. We do a similar thing in Hue.

Reply:
Why did we not go with dispatch update in this case?

Reply:
Ring is a complicated beast. There is 2 endpoints that provide all info about devices and current alerts. Then there are history endpoints per device that are used in this case. But there are multiple entities pulling data from the history endpoint. (last ding, last motion, last activity (which is either ding or motion)

Reply:
It's still not clear to me why a central update and cache with signals telling the entities to update would be worse than this central update with a timed cache where entities pull data, in this situation.

The entities per device that are involved are fixed, except for the camera where only cameras that have subscription are added. So there won't be less or more entities per device pulling data except for the camera case. Maybe the camera entity is the reason?

My point is, that if we know what devices we have and nothing about the features of those devices changes, we should be able to update all devices regardless of platforms.

Reply:
Oh my bad. I thought you asked why not use X, instead of how does it work with the existing dispatch update that Ring uses. 

So I don't have a good reason. And it is indeed kinda complicated now. We now have a health tracker that has an opt-in signal. We could have used the same approach for history. hmm.

Reply:
Want to rewrite it ?

Reply:
I think it's ok like this. I mostly want to understand the reasons for the different designs so I know when to recommend what approach. And I like as few choices as possible for my recommendations.

Reply:
Sorry about my last comment tone. I meant to say want ME to rewrite it. 

I thought about it more and I'll do one more pass cleaning up data fetching. I can do better here.